File: tests/test_distance_metrics.cpp

#include "CLUEstering/CLUEstering.hpp"
#include <cmath>

#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"

TEST_CASE("Test euclidian metric") {
  auto metric = clue::metrics::Euclidean<2>();
  std::array<float, 3> point1{1.f, 2.f, 0.f};
  std::array<float, 3> point2{4.f, 6.f, 0.f};
  CHECK(metric(point1, point2) == doctest::Approx(5.f));
}

TEST_CASE("Test weighted euclidian metric") {
  SUBCASE("Array constructor") {
    auto metric_weights = std::array<float, 2>{1.f, 2.f};
    auto metric = clue::metrics::WeightedEuclidean<2>(metric_weights);
    std::array<float, 3> point1{1.f, 2.f, 0.f};
    std::array<float, 3> point2{4.f, 6.f, 0.f};
    CHECK(metric(point1, point2) == doctest::Approx(std::sqrt(41.f)));
  }

  SUBCASE("Variadic values constructor") {
    auto metric = clue::metrics::WeightedEuclidean<2>(1.f, 2.f);
    std::array<float, 3> point1{1.f, 2.f, 0.f};
    std::array<float, 3> point2{4.f, 6.f, 0.f};
    CHECK(metric(point1, point2) == doctest::Approx(std::sqrt(41.f)));
  }
}

TEST_CASE("Test periodic euclidian metric") {}

TEST_CASE("Test manhattan metric") {
  auto metric = clue::metrics::Manhattan<2>();
  std::array<float, 3> point1{-1.f, 3.f, 0.f};
  std::array<float, 3> point2{0.f, 6.f, 0.f};
  CHECK(metric(point1, point2) == doctest::Approx(4.f));
}

TEST_CASE("Test chebyshev metric") {
  auto metric = clue::metrics::Chebyshev<2>();
  std::array<float, 3> point1{-1.f, 4.f, 0.f};
  std::array<float, 3> point2{0.f, 6.f, 0.f};
  CHECK(metric(point1, point2) == doctest::Approx(2.f));
}

TEST_CASE("Test weighted chebyshev metric") {
  SUBCASE("Array constructor") {
    auto metric_weights = std::array<float, 2>{1.f, 2.f};
    auto metric = clue::metrics::WeightedChebyshev<2>(metric_weights);
    std::array<float, 3> point1{-1.f, 4.f, 0.f};
    std::array<float, 3> point2{0.f, 6.f, 0.f};
    CHECK(metric(point1, point2) == doctest::Approx(4.f));
  }

  SUBCASE("Variadic values constructor") {
    auto metric = clue::metrics::WeightedChebyshev<2>(1.f, 2.f);
    std::array<float, 3> point1{-1.f, 4.f, 0.f};
    std::array<float, 3> point2{0.f, 6.f, 0.f};
    CHECK(metric(point1, point2) == doctest::Approx(4.f));
  }
}

File: tests/test_distance_metrics.py
'''
Testing the distance metrics
'''

import os
import sys
import pandas as pd
import pytest
from sklearn.metrics import silhouette_score
sys.path.insert(1, '../CLUEstering/')
import CLUEstering as clue


@pytest.fixture
def sissa_1000():
    '''
    Returns the dataframe containing the sissa dataset 1000 points
    '''
    return pd.read_csv("../data/sissa_1000.csv")


def test_euclidean(sissa_1000):
    '''
    Tests the euclidean metric
    '''

    c = clue.clusterer(20., 10., 20.)
    c.read_data(sissa_1000)
    assert c.n_dim == 2
    c.choose_metric('euclidean')
    c.run_clue()

    mask = c.cluster_ids != -1
    assert silhouette_score(c.coords.T[mask], c.cluster_ids[mask]) > 0.5


def test_weighted_euclidean(sissa_1000):
    '''
    Tests the weighted euclidean metric
    '''

    c = clue.clusterer(20., 10., 20.)
    c.read_data(sissa_1000)
    assert c.n_dim == 2
    c.choose_metric('weighted_euclidean', [0.5, 0.5])
    c.run_clue()

    mask = c.cluster_ids != -1
    assert silhouette_score(c.coords.T[mask], c.cluster_ids[mask]) > 0.5


def test_manhattan(sissa_1000):
    '''
    Tests the manhattan metric
    '''

    c = clue.clusterer(20., 10., 20.)
    c.read_data(sissa_1000)
    assert c.n_dim == 2
    c.choose_metric('manhattan')
    c.run_clue()

    mask = c.cluster_ids != -1
    assert silhouette_score(c.coords.T[mask], c.cluster_ids[mask]) > 0.5


def test_chebyshev(sissa_1000):
    '''
    Tests the chebyshev metric
    '''

    c = clue.clusterer(20., 10., 20.)
    c.read_data(sissa_1000)
    assert c.n_dim == 2
    c.choose_metric('chebyshev')
    c.run_clue()

    mask = c.cluster_ids != -1
    assert silhouette_score(c.coords.T[mask], c.cluster_ids[mask]) > 0.5

def test_weighted_chebyshev(sissa_1000):
    '''
    Tests the weighted chebyshev metric
    '''

    c = clue.clusterer(20., 10., 20.)
    c.read_data(sissa_1000)
    assert c.n_dim == 2
    c.choose_metric('weighted_chebyshev', [0.5, 0.5])
    c.run_clue()

    mask = c.cluster_ids != -1
    assert silhouette_score(c.coords.T[mask], c.cluster_ids[mask]) > 0.5
